File: erp_bot/src/oip/modules/conversation/application/knowledge_honesty_signoff_policy.py
"""PR-B5 / ADR_0088 — knowledge honesty sign-off (GAP-P2-008 stays REDUCED)."""

from __future__ import annotations

import json
from functools import lru_cache
from pathlib import Path
from typing import Any, Mapping

AUTHORITY = "ADR_0088"
STEP = "PR-B5"
DECISION = "KNOWLEDGE_HONESTY_SIGNOFF_PACK"
REGISTER_GAP_STATUS = "REDUCED"
RUNTIME_GAP_STATUS = "OPEN"
ENGINEERING_GATE_STATUS = "PASS"
LAUNCH_ASK_SIGN_STATUS = "ENGINEERING_PASS"


def _repo_root() -> Path:
    return Path(__file__).resolve().parents[6]


def registry_path() -> Path:
    return (
        _repo_root()
        / "docs"
        / "mokxya-ai"
        / "MAI_KNOWLEDGE_HONESTY_SIGNOFF_REGISTRY.json"
    )


def run_status_path() -> Path:
    return _repo_root() / "artifacts" / "prod-ready-pr-b5" / "RUN_STATUS.json"


@lru_cache(maxsize=1)
def load_knowledge_honesty_signoff_registry() -> dict[str, Any]:
    data = json.loads(registry_path().read_text(encoding="utf-8"))
    if data.get("authority") != AUTHORITY:
        raise RuntimeError("KNOWLEDGE_SIGNOFF_REGISTRY_AUTHORITY_MISMATCH")
    if data.get("decision") != DECISION:
        raise RuntimeError("KNOWLEDGE_SIGNOFF_DECISION_MISMATCH")
    return data


def load_run_status() -> dict[str, Any]:
    return json.loads(run_status_path().read_text(encoding="utf-8"))


def knowledge_honesty_signoff_observability() -> dict[str, Any]:
    reg = load_knowledge_honesty_signoff_registry()
    run = load_run_status()
    return {
        "knowledge_honesty_signoff_step": STEP,
        "knowledge_honesty_signoff_adr": AUTHORITY,
        "knowledge_honesty_signoff_decision": reg["decision"],
        "engineering_gate_status": ENGINEERING_GATE_STATUS,
        "launch_ask_sign_status": LAUNCH_ASK_SIGN_STATUS,
        "staging_professional_attested": bool(run.get("staging_professional_attested")),
        "blocking_fix_required": bool(run.get("blocking_fix_required")),
        "gap_p2_008_register_status": REGISTER_GAP_STATUS,
        "gap_p2_008_status": RUNTIME_GAP_STATUS,
        "gap_p2_008_closed": False,
        "engineering_pack_ready": bool(run.get("engineering_pack_ready")),
        "production_approved": False,
        "is_execution_authority": False,
    }


def assert_knowledge_honesty_signoff_honesty(
    claim: Mapping[str, Any] | None = None,
) -> None:
    reg = load_knowledge_honesty_signoff_registry()
    run = load_run_status()
    honesty = reg.get("honesty") or {}
    if honesty.get("production_approved") is True or run.get("production_approved") is True:
        raise RuntimeError("KNOWLEDGE_SIGNOFF_PRODUCTION_APPROVED")
    if honesty.get("gap_p2_008_closed") is True or run.get("gap_p2_008_closed") is True:
        raise RuntimeError("GAP_P2_008_FALSE_CLOSED")
    if honesty.get("claims_verified") is True:
        raise RuntimeError("CLAIMS_VERIFIED_FORBIDDEN")
    if honesty.get("legal_effective_dates_proven") is True:
        raise RuntimeError("LEGAL_DATES_PROVEN_FORBIDDEN")
    if reg["gap_p2_008"].get("closed") is True:
        raise RuntimeError("GAP_P2_008_FALSE_CLOSED")
    if honesty.get("staging_professional_attested") is True and not run.get(
        "staging_professional_attested"
    ):
        raise RuntimeError("STAGING_PROFESSIONAL_FALSE_ATTESTATION")
    if not claim:
        return
    if claim.get("production_approved") is True:
        raise RuntimeError("KNOWLEDGE_SIGNOFF_PRODUCTION_APPROVED")
    if claim.get("gap_p2_008_closed") is True:
        raise RuntimeError("GAP_P2_008_FALSE_CLOSED")
    if claim.get("claims_verified") is True:
        raise RuntimeError("CLAIMS_VERIFIED_FORBIDDEN")
    if claim.get("legal_effective_dates_proven") is True:
        raise RuntimeError("LEGAL_DATES_PROVEN_FORBIDDEN")
    if claim.get("staging_professional_attested") is True and not run.get(
        "staging_professional_attested"
    ):
        raise RuntimeError("STAGING_PROFESSIONAL_FALSE_ATTESTATION")

File: src/platform/knowledge/knowledgeHonestySignoffPolicy.ts
/**
 * PR-B5 / ADR_0088 — knowledge honesty sign-off (GAP-P2-008 stays REDUCED).
 */

export const KNOWLEDGE_HONESTY_SIGNOFF_ADR = "ADR_0088" as const;
export const KNOWLEDGE_HONESTY_SIGNOFF_STEP = "PR-B5" as const;
export const KNOWLEDGE_HONESTY_SIGNOFF_DECISION =
  "KNOWLEDGE_HONESTY_SIGNOFF_PACK" as const;
export const ENGINEERING_GATE_STATUS = "PASS" as const;
export const LAUNCH_ASK_SIGN_STATUS = "ENGINEERING_PASS" as const;
export const STAGING_PROFESSIONAL_ATTESTED = false;
export const GAP_P2_008_REGISTER_STATUS = "REDUCED" as const;
export const GAP_P2_008_CLOSED = false;
export const PRODUCTION_APPROVED = false;
export const CLAIMS_VERIFIED = false;
export const LEGAL_EFFECTIVE_DATES_PROVEN = false;

export function knowledgeHonestySignoffSnapshot() {
  return {
    authority: KNOWLEDGE_HONESTY_SIGNOFF_ADR,
    step: KNOWLEDGE_HONESTY_SIGNOFF_STEP,
    decision: KNOWLEDGE_HONESTY_SIGNOFF_DECISION,
    engineeringGateStatus: ENGINEERING_GATE_STATUS,
    launchAskSignStatus: LAUNCH_ASK_SIGN_STATUS,
    stagingProfessionalAttested: STAGING_PROFESSIONAL_ATTESTED,
    gapP2008RegisterStatus: GAP_P2_008_REGISTER_STATUS,
    gapP2008Closed: GAP_P2_008_CLOSED,
    claimsVerified: CLAIMS_VERIFIED,
    legalEffectiveDatesProven: LEGAL_EFFECTIVE_DATES_PROVEN,
    productionApproved: PRODUCTION_APPROVED,
  };
}
